HCRC-109 | fix(askem): fix Askem looping to fetch rns.js on article pages

karisal-anders opened the pull request:
## Description

### fix(askem): fix Askem looping to fetch rns.js on article pages

NOTE:
 - There seems to be an underlying issue here that is not fully fixed
   by this commit. This commit fixes the infinite looping to fetch
   Askem's rns.js script and the unnecessary updating of AskemInstance
   but  it's a tip of the iceberg so to speak, there's still something
   being rerendered underneath. So, more work is needed to fix the
   underlying issue.

parts:
 - don't render AskemProvider and AskemFeedbackContainer on server-side
 - don't update used AskemInstance except when it really changes, use
   deep equality to compare

also:
 - add more fields and documentation to Askem's data using
   https://docs.reactandshare.com/
 - pass all the Askem fields to window.rnsData
 - fix AskemContext and AskemProvider default import renaming typos

refs HCRC-106

### refactor: improve Askem api key api config

### refactor: move Matomo, Askem and HDS-style handling of the BaseApp to new files

HCRC-106

## Issues

### Closes

[HCRC-109](https://helsinkisolutionoffice.atlassian.net/browse/HCRC-109)

### Related

## Testing

- checkout PR's branch locally because review environments don't have Askem enabled
- cd apps/sports-helsinki
- yarn
- yarn dev
- open page https://localhost:3000/fi/articles/yleinen/testiartikkeli-syyskuu in browser
- make sure Askem cookies are allowed, if you're not sure clear localhost:3000's cookies or at least the `city-of-helsinki-cookie-consents` cookie
- do twice in a row:
  - scroll down to the bottom of the page
  - scroll back up to the top of the page
  - reload

What is expected:
 - Everything should be fine and the network shouldn't be loading anything in a seemingly infinite loop.
 - Askem functionality should also work all around correctly.

### Automated tests

### Manual testing

Tested locally and couldn't get the infinite network loop fetching Askem's rns.js script to happen anymore.

## Screenshots

## Additional notes


[HCRC-106]: https://helsinkisolutionoffice.atlassian.net/browse/HCRC-106?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[HCRC-109]: https://helsinkisolutionoffice.atlassian.net/browse/HCRC-109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr529.dev.hel.ninja :laughing::tada::tada::tada:

nikomakela commented on packages/components/src/app/BaseApp.tsx:
```diff
@@ -126,14 +139,15 @@ function BaseApp({
-      }),
-    [askemFeedbackConfiguration, askemConsentGiven]
-  );
+  const newAskemConfiguration: AskemConfigs = {
+    ...askemConfigurationInput,
+    consentGiven: askemConsentGiven,
+  };
+
+  if (!askemInstance || !isEqual(askemConfiguration, newAskemConfiguration)) {
+    setAskemConfiguration(newAskemConfiguration);
+    setAskemInstance(createAskemInstance(newAskemConfiguration));
+  }
```
I don't get this. Handling this wit a state should not make a difference compared to a memo. The memo is just a lot cleaner. Both are executed when the dependency array changes.

I need to test this. How was the issue reproducable?

nikomakela replied:
Found the instructions from the PR description 😝 

```
Testing
checkout PR's branch locally because review environments don't have Askem enabled
cd apps/sports-helsinki
yarn
yarn dev
open page https://localhost:3000/fi/articles/yleinen/testiartikkeli-syyskuu in browser
make sure Askem cookies are allowed, if you're not sure clear localhost:3000's cookies or at least the city-of-helsinki-cookie-consents cookie
do twice in a row:
scroll down to the bottom of the page
scroll back up to the top of the page
reload
What is expected:

Everything should be fine and the network shouldn't be loading anything in a seemingly infinite loop.
Askem functionality should also work all around correctly.
```

karisal-anders replied:
useMemo with any object variables is relatively useless IMO as it doesn't compare the objects' contents but simply the references themselves → useMemo always changes the return value if any of useMemo's dependency list's objects' references change even though their contents stay the same. I added deep equality checking to get around this.

nikomakela replied:
It actually can compare deeply: the dependency array could be something like `[JSON.stringify(object)]`. As a reference I found this for example: https://javascript.plainenglish.io/a-less-known-trick-with-react-hook-dependencies-8dafaca7a150

karisal-anders replied:
According to [useMemo documentation](https://react.dev/reference/react/useMemo#memoizing-a-dependency-of-another-hook) it seems they're suggesting to memoize objects used as useMemo dependencies → recursively going back to only using primitive types as dependencies in useMemo in the leaf nodes of the useMemo calls.

The [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) has some limitations:
- BigInt -> throw
- undefined -> null
- Function -> nothing or null
- Symbol -> nothing or null
- NaN -> null
- Infinity -> null
- Map -> "{}"
- Set -> "{}"

which could possibly be overcome with the use of the [replacer ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter)parameter.

nikomakela reviewed (changes requested):
### Test results

❌ I could still reproduce the issue locally ❗ 

![image](https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/9b3128d2-5ece-49ad-8432-f9fc90995da6)

<img width="1780" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/26b29af4-47dd-4d8f-9022-5bfaff501038">


This actually happens even when the askem is disabled, so maybe it's not related to Askem at all 🤔 :

<img width="1771" alt="image" src="https://github.com/City-of-Helsinki/events-helsinki-monorepo/assets/389204/50338463-e932-4f12-86e0-fcf951feb60c">

nikomakela commented:
**Okay, I'm 99% sure, that the problem is in the LayoutCards-module (that is also in the example page of this PR)**

When I exclude all the `LayoutCard` modules, the example page renders OK

```

const NextCmsArticle: NextPage<{
  article: ArticleType;
  breadcrumbs: Breadcrumb[] | null;
  collections: CollectionType[];
}> = ({ article, breadcrumbs, collections }) => {
  const {
    utils: { getRoutedInternalHref },
  } = useConfig();

  const { t: commonTranslation } = useCommonTranslation();
  const { t: appTranslation } = useAppSportsTranslation();
  const { footerMenu } = useContext(NavigationContext);
  console.log('ARTICLE', { article });

  // FIXME: Return null to fix SSR rendering for notFound-page.
  // This is needed only with fallback: true, but should not be needed at all.
  if (!article) return null;

  return (
    <MatomoWrapper>
      <RHHCPage
        className="article-page"
        navigation={<Navigation page={article} />}
        content={
          <>
            <RouteMeta origin={AppConfig.origin} page={article} />
            <RHHCPageContent
              page={
                {
                  ...article,
                  modules: [
                    ...article.modules?.filter(
                      (a) => a?.__typename !== 'LayoutCards'
                    ),
                  ],
                } as ArticleType
              }
              heroContainer={<KorosWrapper />}
              breadcrumbs={
                breadcrumbs && breadcrumbs.length > 0 ? breadcrumbs : undefined
              }
              shareLinks={
                <ShareLinks title={commonTranslation('common:share.article')} />
              }
              collections={
                collections
                  ? cmsHelper.getDefaultCollections(
                      article,
                      getRoutedInternalHref
                    )
                  : []
              }
            />
          </>
        }
        footer={
          <FooterSection
            menu={footerMenu}
            appName={appTranslation('appSports:appName')}
          />
        }
      />
    </MatomoWrapper>
  );
};

```

nikomakela commented:
There are 2 commits that I call improvements in #532 ! 😌

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Events-Helsinki branch is deployed to platta: https://tapahtumat-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**Hobbies-Helsinki branch is deployed to platta: https://harrastukset-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://tapahtumat-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
## TestCafe result is __success__ for https://harrastukset-pr529.dev.hel.ninja :laughing::tada::tada::tada:

terovirtanen commented:
**Sports-Helsinki branch is deployed to platta: https://liikunta-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**VENUE GRAPHQL PROXY is deployed to platta: https://venue-graphql-proxy-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
**EVENTS GRAPHQL PROXY is deployed to platta: https://events-graphql-proxy-pr529.dev.hel.ninja** :rocket::rocket::rocket:

terovirtanen commented:
## TestCafe result is __success__ for https://liikunta-pr529.dev.hel.ninja :laughing::tada::tada::tada:

karisal-anders commented:
> There are 2 commits that I call improvements in #532 ! 😌
> 
> I think that the changes in this PR could be taken in action, but it just does not fix what it is saying to fix.

The commits from #532 are merged into this PR.

Changed the PR's name and the first commit's commit title to reflect what is being done:
- `fix(askem): fix Askem looping to fetch rns.js on article pages`

Also added a note to the first commit's commit message:
```
NOTE:
 - There seems to be an underlying issue here that is not fully fixed
   by this commit. This commit fixes the infinite looping to fetch
   Askem's rns.js script and the unnecessary updating of AskemInstance
   but  it's a tip of the iceberg so to speak, there's still something
   being rerendered underneath. So, more work is needed to fix the
   underlying issue.
```
